Fix spread move effect order

MacChaeger opened the pull request:
TODO: 

- [x] fix the 74 failing tests
- [x] remove the old `moveHit` function and replace it with `spreadMoveHit` (probably should rename `spreadMoveHit` to `moveHit`)
- [x] rewrite `tryMoveHit` to use new functions
- [x] add tests for things
- [x] old gens...

MacChaeger commented:
So yeah, if anyone has any idea why `this` would be undefined in a function where `this` is a `Battle` and the function was called _from that battle_, lemme know.

Zarel commented:
This pull request **introduces 2 alerts** when merging 0a5133e88734ccfea82c8239d5afd6a285a72cc9 into 51cb39a599d0f9b3dfe12087df321d3858a0d19b - [view on LGTM.com](https://lgtm.com/projects/g/Zarel/Pokemon-Showdown/rev/pr-aefc12417a13033c4feef760285e7a5ccc18d9b5)

**new alerts:**

* 2 for Useless assignment to local variable

---

*Comment posted by [LGTM.com](https://lgtm.com)*

Zarel commented:
Which part is the problem?

Zarel commented:
Also you should probably ignore the LGTM useless assignment alerts, I think they're buggy at the moment.

Zarel commented:
I'm pretty sure the assignments you removed were actually important.

MacChaeger commented on data/scripts.js:
```diff
@@ -292,11 +255,290 @@ let BattleScripts = {
+		return moveResult;
+	},
+	tryImmunityEvent(targets, pokemon, move) {
+		let hitResults = this.runEvent('TryImmunity', targets, pokemon, move);
```
It crashes here, claiming that `this` is undefined. I have no idea how that’s possible, since you can’t even call this function without a defined `this`.

MacChaeger replied:
...oh, no, you can. I just realized it now.

MacChaeger replied:
The problem is the way I called the functions from the array.

Zarel replied:
What does crash mean?

An actual uncaught exception? Do you have the stack trace? The stack trace should make it obvious what's going on.

MacChaeger commented on data/scripts.js:
```diff
@@ -292,11 +255,290 @@ let BattleScripts = {
+		let finalResult;
+		let hitResults;
+		for (const step of moveSteps) {
+			hitResults = step(targets, pokemon, move);
```
The problem is this line; the `this` arg isn't passed this way.

MacChaeger commented on data/scripts.js:
```diff
@@ -292,11 +255,291 @@ let BattleScripts = {
+		return hitResults;
+	},
+	tryHitEvent(targets, pokemon, move) {
+		let hitResults = this.runEvent('TryHit', targets, pokemon, move);
```
Ok, new problem: this is somehow not passing an array.

Zarel replied:
"not passing an array"?

MacChaeger replied:
To test this, I put a `console.log('' + target);` inside `runEvent` and it was almost always a single Pokemon despite TypeScript being satisfied that this is always an array.

Zarel replied:
Oh, um, I don't think an array is a valid target for `runEvent`. It only takes single pokemon, sides, or battles, right?

Zarel replied:
Yeah, `target` must be `Pokemon | Side | Battle | null`. `targets` is an array, right? `Pokemon[]`? That's not a valid event target.

Zarel replied:
In theory TypeScript should be yelling at you for that.

MacChaeger replied:
I modified `runEvent` to accept and handle arrays, but it appears to not even be receiving them.

MacChaeger replied:
Ok, I figured it out. The problem is that `Pokemon.getMoveTargets`, which is supposed to only return a `Pokemon` array, isn't returning an array.
`
let targets = pokemon.getMoveTargets(move, target);
		console.log('1. ' + targets instanceof Array);
`
^This prints `false`.

MacChaeger replied:
...because order of operations was wrong. Adding some parens makes it print `1. true`.

Zarel replied:
Why did you modify `runEvent` to handle arrays? Is there a reason why you can't run each target individually?

MacChaeger replied:
Because the `TryHit` event needs to proc its effects in an order that respects `onTryHitPriority` even with multiple targets. That is, Protect always triggers before Lightning Rod, instead of just going left-to-right like how we'd have it, so we need to gather _all_ event handlers from all the targets and sort them together. Additionally, the `AfterMoveSecondary` event even respects speed order instead of the left-to-right order that all the other events use.

Zarel replied:
Oh. :(

Zarel commented:
Oh yeah, that's pretty easy to fix by just never using function pointers as callbacks.

Zarel commented on data/scripts.js:
```diff
@@ -292,11 +255,291 @@ let BattleScripts = {
+		];
+		if (this.gen >= 7) {
+			// Swap step 1 with step 2
+			moveSteps.splice(2, 0, moveSteps.splice(1, 1)[0]);
```
```suggestion
			[moveSteps[1], moveSteps[2]] = [moveSteps[2], moveSteps[1]];
```

is the most readable way to do a swap.

Zarel commented on data/moves.js:
```diff
@@ -1679,7 +1679,9 @@ let BattleMovedex = {
-				return false;
+				if (!move.spreadHit) this.attrLastMove('[miss]');
+				this.add('-miss', source, target);
+				return null;
```
I always get concerned when this sort of thing gets less DRY. What's changing here and why? Was it not a miss message in the past?

MacChaeger replied:
Well, I hate to concern you more, but I don't exactly remember why I did this. I've been working on this for a while, and I've honestly forgotten what some stuff does. Looking at what's written now, I _think_ this might not be necesary?

Zarel replied:
:[

Zarel replied:
You know I'm not going to merge this unless you figure it out.

And ideally write a better solution.

MacChaeger replied:
Yeah, turned out this was unnecessary.

Zarel commented on data/scripts.js:
```diff
@@ -292,11 +255,291 @@ let BattleScripts = {
+				if (!hitResults[i]) {
+					// hit resolved for this target.
+					// remove target and corresponding hitResults entry from their respective arrays
+					targets.splice(i, 1);
```
`for..splice` loops are really ugly. You can usually get away with a `filter` loop. Does that work here?

Zarel commented on data/scripts.js:
```diff
@@ -292,11 +255,291 @@ let BattleScripts = {
+		}
+
+		let moveResult = !!targets.length;
+		if (!moveResult && finalResult === this.NOT_FAILURE) pokemon.moveThisTurnResult = null;
```
After reading this code, I believe it's mostly useless. `finalResult` isn't actually used except to check for equality with `this.NOT_FAILURE`, but the code above ensures that `finalResult` will never be `this.NOT_FAILURE`.

MacChaeger replied:
the problem is that it should be storing `this.NOT_FAILURE` sometimes

MacChaeger replied:
...and it does. The `finalResult !== false` condition allows that possibility.

MacChaeger replied:
Well, the conditional is still wrong. I should point out that the reason for `finalResult` is so that we can check for a move (spread move or single target) that only hits a Protect-like, which doesn't count as a failure for Stomping Tantrum, but if it actually fails against even one target, that takes precedence, hence the `truthy/number > false > this.NOT_FAILURE > undefined` hierarchy outlined in the comment. However, since you've pointed out that I'm checking for move failure a different way, I can simplify that heirarchy to `truthy/number/false > this.NOT_FAILURE > undefined`.

Zarel replied:
You should probably just make a boolean `atLeastOneFailure` which you can manually track. That doesn't seem too hard.

Zarel commented on data/scripts.js:
```diff
@@ -292,11 +255,291 @@ let BattleScripts = {
+		for (const step of moveSteps) {
+			hitResults = step.call(this, targets, pokemon, move);
+			if (!hitResults) continue;
+			for (let i = 0; i < targets.length; i++) {
```
Since `finalResult` isn't actually used, this entire loop can be replaced with:

```js
targets = targets.filter((target, i) => hitResults[i]);
```

MacChaeger replied:
When I do it this way, TS yells at me and says it can't identify `hitResults`'s type any more.

Zarel replied:
You could just manually define it.

Zarel replied:
Although it's true, TypeScript can't infer that `filter` is synchronous, so you'll have to explicitly cast it.

Or use the `any` cast; it's up to you.

Zarel replied:
```js
targets = targets.filter((target, i) =>
  /** @type {any} */ (hitResults)[i]
);
```

MacChaeger replied:
Even if I cast it like that, it _still_ says it implicitly has `any` type. Could that be a legitimate bug with TS?

Zarel replied:
That seems wrong. I'd have to see it in VS Code to figure it out, but honestly if a `//ts-ignore` works, I'd just call it a day.

Zarel commented on data/scripts.js:
```diff
@@ -292,11 +255,289 @@ let BattleScripts = {
+		}
+
+		let finalResult;
+		let hitResults;
```
```suggestion
		/** @type {(number | boolean | "" | undefined)[] | undefined} */
		let hitResults;
```

MacChaeger replied:
If I do this, it complains about `hitResults` potentially being undefined even though I already checked for that. It seems that TS just really hates the `filter` function.

MacChaeger replied:
Meh, I'll just `// @ts-ignore`.

Zarel replied:
Yes, that is in fact what I said:

> Although it's true, TypeScript can't infer that `filter` is synchronous, so you'll have to explicitly cast it.

MacChaeger commented:
> I'm pretty sure the assignments you removed were actually important.

They were, but they were also wrong. They should have been `didAnything = didAnything || false` not `didSomething = didSomething || false`.

Zarel commented on data/scripts.js:
```diff
@@ -292,11 +255,283 @@ let BattleScripts = {
+					atLeastOneFailure = true;
+				}
+				return val || val === 0;
+			});
```
Okay, _this_ is actually a useless assignment – I wonder if LGTM would catch this.

You're not actually using the `filter` results, so this should just be a `for...of` loop.

MacChaeger replied:
I'll do you one better and just use `.some`

MacChaeger commented:
Woo! Only failing 157 tests now!

Zarel commented on data/scripts.js:
```diff
@@ -292,11 +255,282 @@ let BattleScripts = {
+			if (!hitResults) continue;
+			// @ts-ignore
+			targets = targets.filter((val, i) => hitResults[i] || hitResults[i] === 0);
+			atLeastOneFailure = atLeastOneFailure || hitResults.some(val => val === false);
```
`null` is also a failure...

MacChaeger replied:
`null` is a non-issue. All the `step` function handle `null` themselves and return `false` instead.

Zarel replied:
I wonder if that should be changed.

Zarel replied:
That sounds like it could be more DRY, after all.

MacChaeger commented:
Things currently broken (according to existing tests):

- [ ] The `AfterMoveSecondary` event doesn't seem to be running, given that Life Orb isn't dealing damage and Color Change isn't working at all.
- [ ] Explosion is KOing the user without dealing any damage.
- [ ] Immunity abilities, i.e. Dry Skin, don't protect the user from damage or apply their healing/stat-boosting effects
- [ ] Self-switching moves don't switch the user out
- [ ] On-contact events don't activate (Mummy, Rocky Helmet, Iron Barbs)
- [ ] Curse doesn't activate, regardless of user's type
- [ ] Air Balloon doesn't pop
- [ ] Counter/Mirror Coat do nothing
- [ ] Stomping Tantrum appears to be backwards, doubling BP when it shouldn't and not doubling when it should

Zarel commented:
This pull request **introduces 1 alert** when merging 045f0310413356e96102c8f4bfdddeec72d5e978 into 76af9369295b7154eefaf2091aab73d5a9ce927f - [view on LGTM.com](https://lgtm.com/projects/g/Zarel/Pokemon-Showdown/rev/pr-34c1d222d846d60a319170a6ab0e338ba78bb04b)

**new alerts:**

* 1 for Useless assignment to local variable

---

*Comment posted by [LGTM.com](https://lgtm.com)*

Zarel commented on data/scripts.js:
```diff
@@ -1028,6 +1701,21 @@ let BattleScripts = {
 		return CHOOSABLE_TARGETS.has(targetType);
 	},
+
+	reduceMoveResults(states) {
```
`reduce` is a bad name which I suspect you just chose because you run the reduce function on the results.

```suggestion
	combineResults(states) {
```

is what I'd call it.

Zarel commented on data/scripts.js:
```diff
@@ -1028,6 +1701,21 @@ let BattleScripts = {
+			} else if (typeof val === 'number' || (val === true || val === null) && !total || total === undefined) {
+				return val;
+			}
+			return total;
```
I'm also not a huge fan of this reducer.

Here's how I'd do it:

First, don't privilege either side of a reducer function. It doesn't matter which side is the total because if your reducer isn't symmetric, it's probably a bug (or if it isn't, it's still less readable).

```js
combineResults(left, right) {
	const NOT_FAILURE = ''; // local so it doesn't need `this` binding
	if (left === NOT_FAILURE) return right;
	if (right === NOT_FAILURE) return left;

	if (typeof left === 'number' && typeof right === 'number') return left + right;
	if (typeof left === 'number') return left;
	if (typeof right === 'number') return right;

	if (left === null || right === null) return null;
	return false;
},
```

Notice that `combineResults` itself _is_ a reducer, rather than _running_ the reducer.

This is so you don't actually need to use `reduce` on single combinations:

```js
// old:
let didAnything = this.reduceMoveResults(damage);
// new:
let didAnything = damage.reduce(this.combineResults);

// old:
didSomething = this.reduceMoveResults([didSomething, false]);
// new:
didSomething = this.combineResults(didSomething, false);
```

MacChaeger replied:
Is there a way to do this using generics? I tried using
`combineResults?: <T extends number | boolean | null | '' | undefined, U extends number | boolean | null | '' | undefined>(this: Battle, left: T, right: U) => T | U`
```
combineResults(left, right) {
	const NOT_FAILURE = '';
	if (typeof left === 'number' && typeof right === 'number') {
		return left + right;
	} else if (typeof left === 'number') {
		return left;
	} else if (typeof right === 'number') {
		return right;
	} else if (left === null || right === null) {
		return null;
	} else if (left === false || right === false) {
		return false;
	} else if (left === NOT_FAILURE || right === NOT_FAILURE) {
		return NOT_FAILURE;
	} else {
		return undefined;
	}
},
```
in an attempt to get the function to play nice with contexts where `null` and `NOT_FAILURE` aren't relevant, but while this seems right to me, I get errors like
```
Type '<T extends number | boolean | "" | null | undefined, U extends number | boolean | "" | null | undefined>(this: Battle, left: T, right: U) => number | false | "" | null | undefined' is not assignable to type '<T extends number | boolean | "" | null | undefined, U extends number | boolean | "" | null | undefined>(this: Battle, left: T, right: U) => T | U'.
  Type 'number | false | "" | null | undefined' is not assignable to type 'T | U'.
    Type 'undefined' is not assignable to type 'T | U'.
```
 and
`Types T and U have no overlap.`
Is there any hope here, or should I give up on generics?

Zarel replied:
`T` and `U` should be the same type.

`combineResults?: <T extends number | boolean | null | '' | undefined, T extends number | boolean | null | '' | undefined>(this: Battle, left: T, right: T) => T`

And then just recast or ts-ignore all the returns.

Zarel replied:
Also, I believe `undefined` is a success state.

Zarel replied:
Otherwise why is `NOT_FAILURE` not `undefined`?

MacChaeger replied:
Well, I got Typescript to stop complaining by going with this idea.
```
combineResults(left, right) {
	const NOT_FAILURE = 'string';
	let resultsPriorities = ['undefined', NOT_FAILURE, 'null', 'false', 'number'];
	if (resultsPriorities.indexOf(typeof left) > resultsPriorities.indexOf(typeof right)) {
		return left;
	} else {
		return right;
	}
},
```
However, this entire approach of using `didSomething = this.combineResults(didSomething, hitResults)` instead of just `didSomething = didSomething || hitResult` has been breaking more tests than they've fixed, even though it _should_ be a better way.

Zarel replied:
Well, your `combineResults` implementation seems wrong to me, which might be part of it. :p

I guess I'll look at your code at some point.

MacChaeger replied:
Yeah, it was wrong. There's just 16 failing tests now!

MacChaeger replied:
`undefined` isn't a success condition here. Or, it wasn't, but now that you mention it, I should probably use `NOT_FAILURE` as the default starting value instead of `undefined` since `undefined` _is_ a success condition sometimes.

Zarel replied:
Oh, it can't. Because `NOT_FAILURE` means "end this chain", while `undefined` means "keep going".

MacChaeger commented:
I feel like this problem is one of these
![image](https://user-images.githubusercontent.com/21160928/53685062-539fd980-3cdb-11e9-8637-ff35cbd5c86c.png)
and I'm trying to solve it by just shaking the thing.

MacChaeger commented:
Only 2 tests remaining!!!!

MacChaeger commented on sim/battle.ts:
```diff
@@ -761,7 +771,24 @@ export class Battle extends Dex.ModdedDex {
+		}
 		for (const handler of handlers) {
+			if (handler.index !== undefined) {
+				if (!targetRelayVars[handler.index]) continue;
```
Counter + Disguise isn't working because `targetRelayVars[handler.index]` in that case is 0, referring to the amount of damage dealt. However, I can't seem to get this code to run in that case and only that case, because everything I've tried breaks at least 5 other tests, including checking the event id and skipping that check the first time through the loop.

MacChaeger commented:
All tests passing! This can actually be merged now, as the other things to do are merely to remove code duplication and give older gens the same update with respect to spread moves.

MacChaeger commented:
Well, nobody wrote me any tests for doubles-specific stuff like Emergency Exit activation order, and there's actually still research to do on that front, so I suppose I should write some tests at least before this is _really_ ready.

MacChaeger commented:
Gah, realized last night in bed that there’s a couple obvious mistakes I made with the Counter and Emergency Exit fixes.

Zarel commented on data/scripts.js:
```diff
@@ -296,158 +259,248 @@ let BattleScripts = {
+			// 0. check for semi invulnerability
+			this.tryImmunityEvent,
+
+			// 1. check for type immunity (this is step 2 in gen 7+)
```
I would start with the Gen 7 order, and swap them in gen 6 and below.

MacChaeger replied:
Ok

MacChaeger commented:
Actually, lemme change how I did gens 2 & 3.

MacChaeger commented:
I think that's everything.

Zarel commented on config/datacenters.csv:
```diff
@@ -1888,6 +1889,7 @@
 162.248.72.0,162.248.79.255,Volume Drive,http://volumedrive.com
 162.248.160.0,162.248.167.255,KW Datacenter CA,http://www.kwdatacentre.com/
 162.250.120.0,162.250.127.255,InterServer.net,http://www.interserver.net/
+162.251.232.0,162.251.239.255,Fiberhub,https://www.fiberhub.com/
```
Botched merge

Zarel commented on build:
```diff
@@ -10,17 +10,31 @@ try {
-shell('npx sucrase -q ./sim -d ./.sim-dist --transforms typescript,imports --enable-legacy-typescript-module-interop');
+sucrase('./sim', './.sim-dist');
+sucrase('./lib', './.lib-dist');
+rewrite('lib', '.lib-dist', '.sim-dist');
```
Botched merge

Zarel commented:
I'm not sure why all of you have merge issues, but I don't.

Is rebasing a problem here?

MacChaeger commented:
All I did was rebase, and then the commit I was introducing was suddenly a merge conflict when it shouldn’t have been, and I had no idea that this had gone this poorly until you commented on it just now. I honestly don’t know what even happened here.

MacChaeger commented:
Hey, I fixed it, I think!

Zarel commented:
I see another conflict...

scheibo commented on sim/battle.ts:
```diff
@@ -1771,6 +1822,115 @@ export class Battle extends Dex.ModdedDex {
+			if (targetDamage !== 0) targetDamage = this.clampIntRange(targetDamage, 1);
+
+			if (this.gen <= 1) {
+				// FIXME: The stored damage should be calculated ignoring Substitute.
```
Remove FIXME after #5211

scheibo commented on sim/battle.ts:
```diff
@@ -1771,6 +1822,115 @@ export class Battle extends Dex.ModdedDex {
+		let retVals: (number | boolean | undefined)[] = [];
+		if (typeof effect === 'string' || !effect) effect = this.getEffect(effect);
+		for (let i = 0; i < damage.length; i++) {
+			let target = targetArray[i];
```
const?

scheibo commented on sim/battle.ts:
```diff
@@ -1771,6 +1822,115 @@ export class Battle extends Dex.ModdedDex {
+			if (targetDamage !== 0) target.hurtThisTurn = true;
+			if (source && effect.effectType === 'Move') source.lastDamage = targetDamage;
+
+			let name = effect.fullname;
```
`const name = effect.fullname === 'tox' ? 'psn' : effect.fullname;`?

scheibo commented on sim/battle.ts:
```diff
@@ -1771,6 +1822,115 @@ export class Battle extends Dex.ModdedDex {
+		if (instafaint) {
+			for (let i = 0; i < targetArray.length; i++) {
+				if (!retVals[i]) continue;
+				let target = targetArray[i];
```
const?

scheibo commented on sim/battle.ts:
```diff
@@ -1771,6 +1822,115 @@ export class Battle extends Dex.ModdedDex {
+		// @ts-ignore TODO: AfterDamage passes an Effect, not an ActiveMove
+		if (!effect.flags) effect.flags = {};
+		if (instafaint) {
+			for (let i = 0; i < targetArray.length; i++) {
```
for..of?

scheibo commented on sim/battle.ts:
```diff
@@ -3196,10 +3357,100 @@ export class Battle extends Dex.ModdedDex {
+	}
+
+	spreadMoveHit(
+		targets: (Pokemon | false | null)[], pokemon: Pokemon, move: ActiveMove, moveData?: ActiveMove,
```
maybe use `type` aliases for ` (Pokemon | false | null)[]` and `(number | undefined | boolean)[]` if you can figure out a name for them?

scheibo commented on data/scripts.js:
```diff
@@ -296,158 +259,252 @@ let BattleScripts = {
+			targets = targets.filter((val, i) => hitResults[i] || hitResults[i] === 0);
+			atLeastOneFailure = atLeastOneFailure || hitResults.some(val => val === false);
+			if (!targets.length) {
+				// console.log(step.name);
```
remove commented out log and just make this one line?

Zarel replied:
It's a useful breakpoint point; I'm fine with leaving it in.

scheibo commented on data/scripts.js:
```diff
@@ -296,158 +259,252 @@ let BattleScripts = {
-		hitResult = this.runEvent('TryImmunity', target, pokemon, move);
-		if (!hitResult) {
-			if (hitResult !== null) {
+		let moveResult = !!targets.length;
```
const?

scheibo commented on data/scripts.js:
```diff
@@ -296,158 +259,252 @@ let BattleScripts = {
+		return moveResult;
+	},
+	tryImmunityEvent(targets, pokemon, move) {
+		let hitResults = this.runEvent('TryImmunity', targets, pokemon, move);
```
const?

scheibo commented on data/scripts.js:
```diff
@@ -296,158 +259,252 @@ let BattleScripts = {
 
-		if (this.gen < 7 && (!move.ignoreImmunity || (move.ignoreImmunity !== true && !move.ignoreImmunity[move.type])) && !target.runImmunity(move.type, true)) {
-			return false;
+		let hitResults = [];
```
const?

scheibo commented on data/scripts.js:
```diff
@@ -296,158 +259,252 @@ let BattleScripts = {
+		return hitResults;
+	},
+	tryHitEvent(targets, pokemon, move) {
+		let hitResults = this.runEvent('TryHit', targets, pokemon, move);
```
const?

scheibo commented on data/scripts.js:
```diff
@@ -296,158 +259,252 @@ let BattleScripts = {
+		return hitResults;
+	},
+	powderImmunity(targets, pokemon, move) {
+		let hitResults = [];
```
const?

scheibo commented on data/scripts.js:
```diff
@@ -296,158 +259,252 @@ let BattleScripts = {
-			if (!target.illusion) this.hint("In Gen 7, Dark is immune to Prankster moves.");
-			this.add('-immune', target);
-			return false;
+		for (let [i, target] of targets.entries()) {
```
Oh hey, here we're using `entries` :). Can you `const` with `entries` (everywhere in this file)?

scheibo commented on data/scripts.js:
```diff
@@ -296,158 +259,252 @@ let BattleScripts = {
+		return hitResults;
+	},
+	pranksterImmunity(targets, pokemon, move) {
+		let hitResults = [];
```
const?

scheibo commented on data/scripts.js:
```diff
@@ -296,158 +259,252 @@ let BattleScripts = {
+		return hitResults;
+	},
+	accuracy(targets, pokemon, move) {
+		let hitResults = [];
```
const?

scheibo commented on data/scripts.js:
```diff
@@ -296,158 +259,252 @@ let BattleScripts = {
-					this.add('-immune', target, '[ohko]');
-					return false;
+			} else {
+				let boostTable = [1, 4 / 3, 5 / 3, 2, 7 / 3, 8 / 3, 3];
```
const?

scheibo commented on data/scripts.js:
```diff
@@ -472,85 +529,142 @@ let BattleScripts = {
+	},
+	moveHitLoop(targets, pokemon, move) { // temp name
+		/** @type {(number | boolean | undefined)[]} */
+		let damage = [];
```
const?

scheibo commented on data/scripts.js:
```diff
@@ -472,85 +529,142 @@ let BattleScripts = {
+			// like this (Triple Kick)
+			if (target && move.multiaccuracy && i > 0) {
+				let accuracy = move.accuracy;
+				let boostTable = [1, 4 / 3, 5 / 3, 2, 7 / 3, 8 / 3, 3];
```
seems like there should be a canonical place for accuracy and stat boost tables.

scheibo commented on data/scripts.js:
```diff
@@ -472,85 +529,142 @@ let BattleScripts = {
+				let boostTable = [1, 4 / 3, 5 / 3, 2, 7 / 3, 8 / 3, 3];
+				if (accuracy !== true) {
+					if (!move.ignoreAccuracy) {
+						let boosts = this.runEvent('ModifyBoost', pokemon, null, null, Object.assign({}, pokemon.boosts));
```
const for these 2, and below?

scheibo commented on data/scripts.js:
```diff
@@ -472,85 +529,142 @@ let BattleScripts = {
-				moveDamage = this.moveHit(target, pokemon, move);
-				if (moveDamage === false) break;
-				if (nullDamage && (moveDamage || moveDamage === 0 || moveDamage === undefined)) nullDamage = false;
+			let moveData = move;
```
const?

scheibo commented on data/scripts.js:
```diff
@@ -561,63 +675,39 @@ let BattleScripts = {
+	spreadMoveHit(targets, pokemon, moveOrMoveName, moveData, isSecondary, isSelf) {
+		// Hardcoded for single-target purposes
+		// (no spread moves have any kind of onTryHit handler)
+		let target = targets[0];
```
const these 2?

scheibo commented on data/scripts.js:
```diff
@@ -670,84 +808,99 @@ let BattleScripts = {
+		// @ts-ignore
+		let didAnything = damage.reduce(this.combineResults, undefined);
+		for (let i = 0; i < targets.length; i++) {
+			let target = targets[i];
```
const?

scheibo commented on data/scripts.js:
```diff
@@ -812,21 +984,27 @@ let BattleScripts = {
+	},
+	forceSwitch(damage, targets, pokemon, move) {
+		for (let i = 0; i < targets.length; i++) {
+			let target = targets[i];
```
const

scheibo commented on data/scripts.js:
```diff
@@ -1028,6 +1206,19 @@ let BattleScripts = {
+	combineResults(left, right) {
+		const NOT_FAILURE = 'string';
+		const NULL = 'object';
+		let resultsPriorities = ['undefined', NOT_FAILURE, NULL, 'boolean', 'number'];
```
const

MacChaeger commented on sim/battle.ts:
```diff
@@ -872,8 +910,21 @@ export class Battle extends Dex.ModdedDex {
+				const pokemon = thing[i];
+				// console.log(`Event: ${eventName}, Target: ${'' + pokemon}, ${i}`);
+				let curHandlers = this.findEventHandlers(pokemon, eventName, sourceThing);
+				curHandlers.forEach(val => {
```
Guess I missed this one.

MacChaeger commented on data/abilities.js:
```diff
@@ -855,7 +855,8 @@ let BattleAbilities = {
 			if (!source || source === target || !target.hp || !move.totalDamage) return;
 			if (target.hp <= target.maxhp / 2 && target.hp + move.totalDamage > target.maxhp / 2) {
-				if (!this.canSwitch(target.side) || target.forceSwitchFlag || target.switchFlag) return;
+				if (!this.canSwitch(target.side) || target.forceSwitchFlag || target.switchFlag || this.ejecting) return;
```
I suppose if I add `this.ejecting = true` to all effects that trigger a switch, we wouldn't need to check `target.forceSwitchFlag` or `target.switchFlag`. @Marty-D would that work?

Zarel replied:
Any reason it's not `this.ejecting = target`?

And if it _is_ `this.ejecting = target`, do we still need all the switch flags?

Zarel replied:
And if `this.ejecting = target` _is_ the better approach, can that refactor be done in a separate PR?

MacChaeger replied:
Still need @Marty-D's explanation of that, since, as I understand it, there are cases where more than one Pokemon can switch out at the same time.

MacChaeger replied:
For now, I'm just going to leave that fix out and redo it in a separate PR once this one's merged, since it won't make sense until then.

MacChaeger commented on data/mods/gen4/scripts.js:
```diff
@@ -89,213 +89,16 @@ let BattleScripts = {
-				this.eachEvent('Update');
+	tryImmunityEvent(targets, pokemon, move) {
+		let hitResults = this.runEvent('TryImmunity', targets, pokemon, move);
+		for (let i = 0; i < targets.length; i++) {
```
Ooh, I didn't know about the `.entries()` thing until now.

MacChaeger commented:
rip, the coauthor didn't work

scheibo commented on dev-tools/globals.ts:
```diff
@@ -732,24 +732,46 @@ interface Format extends Readonly<BasicEffect & FormatsData> {
 	runAction?: (this: Battle, action: Actions["Action"]) => void
 	runMegaEvo?: (this: Battle, pokemon: Pokemon) => boolean
 	runMove?: (this: Battle, move: Move, pokemon: Pokemon, targetLoc: number, sourceEffect?: Effect | null, zMove?: string, externalMove?: boolean) => void
+	runMoveEffects?: (this: Battle, damage:SpreadMoveDamage, targets: SpreadMoveTargets, source: Pokemon, move: ActiveMove, moveData: ActiveMove, isSecondary: boolean | undefined, isSelf: boolean | undefined) => SpreadMoveDamage
```
```suggestion
	runMoveEffects?: (this: Battle, damage: SpreadMoveDamage, targets: SpreadMoveTargets, source: Pokemon, move: ActiveMove, moveData: ActiveMove, isSecondary: boolean | undefined, isSelf: boolean | undefined) => SpreadMoveDamage
```

Zarel commented:
(For the record, I definitely think we need to move most script functions from `scripts` back into `battle`. The distinction is _way_ too arbitrary.

Zarel commented:
Although if we did it before this, that would probably make this one easier to review... decisions, decisions... I guess it can be your call.

MacChaeger commented:
If we move them before this, then the functions we're moving won't exist and I wouldn't put it past you to change your mind and decide that some of these new functions should still be in `scripts.js`. I vote after.

Zarel commented on data/scripts.js:
```diff
@@ -1028,6 +1204,19 @@ let BattleScripts = {
+		} else if (left && !right && right !== 0) {
+			return left;
+		} else {
+			return right;
```
Why aren't we adding `left + right` when they're both numbers?

MacChaeger replied:
Because in all the places this function is used right now, it's not actually possible for them both to be numbers, but I can add that functionality back in now if you want.

Zarel replied:
I think it's worth it, just because of the name.

Zarel replied:
I don't want someone to start using it later and get tripped up by it.

MacChaeger replied:
Oh yeah, now I remember why, because TypeScript doesn't understand that returning a number from this function is OK if both parameters are numbers.

scheibo replied:
I mean, in the worst case there's always `// @ts-ignore`...

Zarel commented on data/scripts.js:
```diff
@@ -1028,6 +1204,22 @@ let BattleScripts = {
 		return CHOOSABLE_TARGETS.has(targetType);
 	},
+
+	// @ts-ignore I can add two numbers and return a number, ok?
```
```suggestion
```

This is a really dangerous place for a `ts-ignore`. I'd just cast the result as `any`.

Zarel commented on data/scripts.js:
```diff
@@ -1028,6 +1204,22 @@ let BattleScripts = {
+		} else if (left && !right && right !== 0) {
+			return left;
+		} else if (typeof left === 'number' && typeof right === 'number') {
+			return left + right;
```
```suggestion
			return /** @type {any} */ (left + right);
```

Zarel commented:
Honestly, though, this is a bit too many changes to easily review.

Zarel commented:
It's too bad your commits aren't neat enough to review them individually. :(

Zarel commented:
Anyway, can you go over why `useMoveInner` was added to Gen 3, and `tryMoveHit` were removed from Gen 4? That seems like the most non-sequitur part of this.

Zarel commented on data/scripts.js:
```diff
@@ -296,165 +259,259 @@ let BattleScripts = {
+		if (targets.length > 1) move.spreadHit = true;
+
+		/** @type {((targets: Pokemon[], pokemon: Pokemon, move: ActiveMove) => (number | boolean | "" | undefined)[] | undefined)[]} */
+		let moveSteps = [
```
I think I would prefix all 8 hit step functions with `hitStep`.

So `hitStepTryImmunity`, `hitStepTryHit`, etc.

Zarel replied:
These function names are kind of vague without the namespacing.

MacChaeger commented:
`useMoveInner` was added to Gen 3 because Gen 3's behavior when it comes to move effect order is too weird to fit into how I implemented `useMoveInner` for gens 4-7, so it's going to be using the old system. That is, the one that doesn't do spread move effect order properly. According to @Marty-D there's really no need for it in Gen 3, anyway. I removed `tryMoveHit` from Gen 4 because the only difference between it's version and Gen 7's was step order, and that's already handled in the Gen 7 version, so the only thing necessary was to override the `tryImmunity` step which had slightly different code in Gen 4's `tryMoveHit`.

Zarel commented:
Also, in hindsight, this is too much code to stuff into Scripts. This isn't just about delaying moving code _out_ of Scripts, you're adding a whole bunch of new code _into_ Scripts, and I think it's best to remove it _first_.

Zarel commented:
It's also easier to review `.ts` code than `.js` code.

Zarel commented:
(The Gen 3 and Gen 4 changes sound good, btw.)

MacChaeger commented:
The only actual new code is `combineResults` and maybe `moveHitLoop`. The rest is just old code moved into new functions and put into loops. That said, if you want functions moved to `battle.ts`, just give me a list and they'll be moved.

Zarel commented:
`combineResults` should definitely be moved.

Zarel commented on sim/battle.ts:
```diff
@@ -872,8 +908,20 @@ export class Battle extends Dex.ModdedDex {
-	findEventHandlers(thing: Pokemon | Side | Battle, eventName: string, sourceThing?: Pokemon | null) {
+	findEventHandlers(thing: Pokemon | Pokemon[] | Side | Battle, eventName: string, sourceThing?: Pokemon | null) {
 		let handlers: AnyObject[] = [];
+		if (thing instanceof Array) {
```
```suggestion
		if (Array.isArray(thing)) {
```

Zarel commented on sim/battle.ts:
```diff
@@ -1771,6 +1819,110 @@ export class Battle extends Dex.ModdedDex {
+
+		return retVals;
+	}
 	damage(
```
Can `damage` just delegate to `spreadMoveDamage`? That might make the diff clearer.

MacChaeger replied:
Oh yeah, that's probably fine.

Zarel commented on sim/battle.ts:
```diff
@@ -851,7 +887,7 @@ export class Battle extends Dex.ModdedDex {
 		this.event = parentEvent;
 
-		return relayVar;
+		return target instanceof Array ? targetRelayVars : relayVar;
```
```suggestion
		return Array.isArray(target) ? targetRelayVars : relayVar;
```

Zarel commented on sim/battle.ts:
```diff
@@ -3176,6 +3248,23 @@ export class Battle extends Dex.ModdedDex {
 		}
 	}
 
+	combineResults<T extends number | boolean | null | '' | undefined, U extends number | boolean | null | '' | undefined>(
```
You probably should define a `HitResult` type.

Or actually I will.

Zarel commented on sim/battle.ts:
```diff
@@ -3176,6 +3248,23 @@ export class Battle extends Dex.ModdedDex {
 	}
 
+	combineResults<T extends number | boolean | null | '' | undefined, U extends number | boolean | null | '' | undefined>(
+		left: T, right: U): T | U {
```
```suggestion
		left: T, right: U
	): T | U {
```

Zarel commented on sim/battle.ts:
```diff
@@ -3196,10 +3285,100 @@ export class Battle extends Dex.ModdedDex {
+
+	spreadMoveHit(
+		targets: SpreadMoveTargets, pokemon: Pokemon, move: ActiveMove, moveData?: ActiveMove,
+		isSecondary?: boolean, isSelf?: boolean): [SpreadMoveDamage, SpreadMoveTargets] {
```
```suggestion
		isSecondary?: boolean, isSelf?: boolean
	): [SpreadMoveDamage, SpreadMoveTargets] {
```

Zarel commented on sim/battle.ts:
```diff
@@ -3196,10 +3285,100 @@ export class Battle extends Dex.ModdedDex {
+
+	tryPrimaryHitEvent(
+		damage: SpreadMoveDamage, targets: SpreadMoveTargets, pokemon: Pokemon, move: ActiveMove,
+		moveData: ActiveMove, isSecondary: boolean | undefined): SpreadMoveDamage {
```
```suggestion
		moveData: ActiveMove, isSecondary: boolean | undefined
	): SpreadMoveDamage {
```

scheibo commented:
@MacChaeger - can you please fix the merge conflicts so this can land? :)
@Zarel - is letting this linger on in code review productive at this point (I agree its too big to review in a way we can be fully confident about), or can we just land it and deal with the fallout?

MacChaeger commented:
Oh dear.
![i got 99 problems oh wait no i don't](https://user-images.githubusercontent.com/21160928/54484277-e474ca00-4830-11e9-8bb6-f3276f05846e.png)

MacChaeger commented:
Fucking stupid github editor

urkerab commented on data/scripts.js:
```diff
@@ -632,32 +722,80 @@ let BattleScripts = {
 				this.attrLastMove('[still]');
 			}
-			return false;
+			return [[false], targets]; // single-target only
```
https://www.smogon.com/forums/posts/8110059

Synchronoise used to use `TryHit` single event to mark the move as immune if the target did not share a type with the user. However this event now only runs for the first Pokémon, which means that it hits no Pokémon at all depending on whether the ally does not share a type, unless it misses the ally due to semi-invulnerability, which is in itself a bug which I forgot to file, as synchronise should never miss an immune target.

Checking other moves, it looks like Captivate is also expecting a `TryHit` single event for each target. I don't know what the on-cart mechanics of Captivate are when only one foe has a suitable gender, or what happens if the unsuitable foe is semi-invulnerable.

Speaking of semi-invulnerability, when I tried to reproduce this, I happened to use Toxic from a Poison type to a semi-invulnerable target. It missed. This is also incorrect behaviour; since Gen 6, this should work like No Guard.

Zarel commented on data/scripts.js:
```diff
@@ -300,165 +263,259 @@ let BattleScripts = {
-			if (hitResult !== null) {
+		const moveResult = !!targets.length;
+		if (!moveResult && !atLeastOneFailure) pokemon.moveThisTurnResult = null;
+		if (move.spreadHit) this.attrLastMove('[spread] ' + targets.join(','));
```
You forgot to use `hitTargets` (technically an array of slots) instead of `targets`. You did it correctly in Gen 3, too.

Zarel commented on sim/battle.ts:
```diff
@@ -447,13 +447,21 @@ export class Battle extends Dex.ModdedDex {
 			((b.priority || 0) - (a.priority || 0)) ||
 			((b.speed || 0) - (a.speed || 0)) ||
 			-((b.subOrder || 0) - (a.subOrder || 0)) ||
+			((a.thing && b.thing) ? -(b.thing.abilityOrder - a.thing.abilityOrder) : 0) ||
```
Do you remember why this was added? It seems to have caused the deterministic speed ties I've been trying to fix:

https://github.com/smogon/pokemon-showdown/pull/6319#issuecomment-579372632
